src/v/cluster/non_replicable_partition.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "cluster/partition.h"
+#include "storage/fwd.h"
+
+#include <seastar/core/gate.hh>
+
+namespace cluster {
+
+class safe_shutdown_appender {
+public:
+    safe_shutdown_appender(
+      ss::gate& g, storage::log_appender&& appender) noexcept;
+
+    ss::future<ss::stop_iteration> operator()(model::record_batch&);
+    ss::future<storage::append_result> end_of_stream();
+
+private:
+    ss::gate& _gate;
+    storage::log_appender _appender;
+};
+
+class non_replicable_partition {
```

Comment:
> his PR attempts to solve these problems by creating an abstraction called cluster::non_replicable_partition. The solution works in the same way cluster::partitions are shut down. This class looks and feels like a normal cluster::partition, however its much simpler.

why are we not creating a partition specialization?

Reply:
Making a subclass of partition would inherit raft internals which don't make sense for materialized partitions.

Reply:
it's hard to know exactly what this is going to to look like in practice, but it seems like it might cause a lot of change to code paths that might use one partition type or the other.

Reply:
i was under the impression that one of the tasks we'd be going through with the new co-proc work was to generalize the partition abstraction to handle the co-proc use case.

Reply:
Yes but I did come to a conclusion about this particular point. I wasn't going to try to make one unified abstraction with the partition interface because the existing one has everything to do with raft. There is already a solution called kafka::partition_proxy which chooses the correct implementation after a lookup in the topics table which determines if the topic is replicable or not

Reply:
but we can't return a non-replicable topic partition through that interface, right? because it isn't a subclass of partition? so will we end up duplicated a bunch of code paths?

Reply:
so the code path isn't duplicated because the interface unifies the abstraction

Reply:
wouldn't there need to be a partition_proxy::impl in this PR?

Reply:
I think the change needed would be having the `materialized_partition` take a `non_replicable_partition` as a value in its constructor. Didn't know if i should do it in this PR or a subsequent one

Reply:
Will make the next changes to include this